lib/Completion/Bridge/TolerantParser/ChainTolerantCompletor.php

```diff
@@ -29,11 +29,18 @@ public function __construct(private array $tolerantCompletors, Parser $parser =
     public function complete(TextDocument $source, ByteOffset $byteOffset): Generator
     {
         $truncatedSource = $this->truncateSource((string) $source, $byteOffset->toInt());
+        $truncatedLength = strlen($truncatedSource);
 
-        $node = $this->parser->parseSourceFile($truncatedSource)->getDescendantNodeAtPosition(
-            // the parser requires the byte offset, not the char offset
-            strlen($truncatedSource)
-        );
+        $node = $this->parser
+            ->parseSourceFile(substr_replace((string) $source, ' ', $truncatedLength, 0))
+            ->getDescendantNodeAtPosition(
+                // the parser requires the byte offset, not the char offset
+                $truncatedLength,
+            );
+
+        if ($node->getEndPosition() > $truncatedLength) {
+            $node = $this->parser->parseSourceFile($truncatedSource)->getDescendantNodeAtPosition($truncatedLength);
+        }
```

Comment:
@dantleech @mamazu  It's hacky and probably worth a better replacement, but passes all current tests. 

- on truncated source cannot determine attribute context (whether it's attached to class, method, etc.) because the relevant information is after the cursor
- without truncation: the data set `does not use variables declared after offset a` in `WorseParameterCompletorTest` (and not only) does not pass. 
- inserting additional space after the cursor before parsing does not break any existing behavior but seems to help to get the correct node if there are no white space after the cursor.